src/core-sdk-tasks/DownloadFile.cs

```diff
@@ -39,31 +54,56 @@ public override bool Execute()
                 var filePath = Uri.Substring(FileUriProtocol.Length);
                 Log.LogMessage($"Copying '{filePath}' to '{DestinationPath}'");
                 File.Copy(filePath, DestinationPath);
+                return true;
             }
-            else
-            {
-                Log.LogMessage(MessageImportance.High, $"Downloading '{Uri}' to '{DestinationPath}'");
 
+            Log.LogMessage(MessageImportance.High, $"Downloading '{Uri}' to '{DestinationPath}'");
+
+            Random rng = new Random();
+            for (int retryNumber = 0; retryNumber < MaxRetries; retryNumber++)
+            {
                 using (var httpClient = new HttpClient())
                 {
-                    var getTask = httpClient.GetStreamAsync(Uri);
-
                     try
                     {
+                        var httpResponse = await httpClient.GetAsync(Uri);
+
+                        if (httpResponse.StatusCode == HttpStatusCode.NotFound)
+                        {
+                            if (!string.IsNullOrEmpty(PrivateUri))
+                            {
+                                httpResponse = await httpClient.GetAsync(PrivateUri);
+                            }
+
+                            // The Azure Storage REST API returns '400 - Bad Request' in some cases
+                            // where the resource is not found on the storage.
+                            // https://docs.microsoft.com/en-us/rest/api/storageservices/common-rest-api-error-codes
+                            if (httpResponse.StatusCode == HttpStatusCode.NotFound ||
+                                httpResponse.ReasonPhrase.IndexOf("The requested URI does not represent any resource on the server.", StringComparison.OrdinalIgnoreCase) >= 0)
+                            {
+                                return false;
+                            }
+                        }
+
                         using (var outStream = File.Create(DestinationPath))
                         {
-                            getTask.Result.CopyTo(outStream);
+                            await httpResponse.Content.CopyToAsync(outStream);
                         }
+
+                        return true;
                     }
-                    catch (Exception)
+                    catch (Exception e)
                     {
+                        var secondaryUri = !string.IsNullOrWhiteSpace(PrivateUri) ? $"/'{PrivateUri}'" : string.Empty;
+                        Log.LogMessage($"Problems downloading file from '{Uri}'{secondaryUri}. {e.Message} {e.StackTrace}");
                         File.Delete(DestinationPath);
-                        throw;
                     }
                 }
+
+                await System.Threading.Tasks.Task.Delay(rng.Next(3000, 10000));
             }
 
-            return true;
+            return false;
```

Comment:
Does any actionable error show up as when it fails this way? Since nothing was actually thrown I'm worried you'd have to look all the way to the binlog to diagnose even the basics.

Reply:
Added a `throw HttpResponseException` here.